ast/src/parsed/mod.rs

```diff
@@ -332,24 +332,97 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
+#[derive(Debug, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
 pub enum Expression<Ref = NamespacedPolynomialReference> {
-    Reference(Ref),
-    PublicReference(String),
+    Reference(SourceRef, Ref),
+    PublicReference(SourceRef, String),
     // A number literal and its type.
-    Number(Number),
-    String(String),
-    Tuple(Vec<Self>),
-    LambdaExpression(LambdaExpression<Self>),
-    ArrayLiteral(ArrayLiteral<Self>),
-    UnaryOperation(UnaryOperation<Self>),
-    BinaryOperation(BinaryOperation<Self>),
-    IndexAccess(IndexAccess<Self>),
-    FunctionCall(FunctionCall<Self>),
-    FreeInput(Box<Self>),
-    MatchExpression(MatchExpression<Self>),
-    IfExpression(IfExpression<Self>),
-    BlockExpression(BlockExpression<Self>),
+    Number(SourceRef, Number),
+    String(SourceRef, String),
+    Tuple(SourceRef, Vec<Self>),
+    LambdaExpression(SourceRef, LambdaExpression<Self>),
+    ArrayLiteral(SourceRef, ArrayLiteral<Self>),
+    UnaryOperation(SourceRef, UnaryOperation<Self>),
+    BinaryOperation(SourceRef, BinaryOperation<Self>),
+    IndexAccess(SourceRef, IndexAccess<Self>),
+    FunctionCall(SourceRef, FunctionCall<Self>),
+    FreeInput(SourceRef, Box<Self>),
+    MatchExpression(SourceRef, MatchExpression<Self>),
+    IfExpression(SourceRef, IfExpression<Self>),
+    BlockExpression(SourceRef, BlockExpression<Self>),
+}
+
+// Comparison function for expressions that ignore source information.
+impl<Ref: PartialEq> PartialEq for Expression<Ref> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Expression::Reference(_, a), Expression::Reference(_, b)) => a == b,
+            (Expression::PublicReference(_, a), Expression::PublicReference(_, b)) => a == b,
+            (Expression::Number(_, a), Expression::Number(_, b)) => a == b,
+            (Expression::String(_, a), Expression::String(_, b)) => a == b,
+            (Expression::Tuple(_, a), Expression::Tuple(_, b)) => a == b,
+            (Expression::LambdaExpression(_, a), Expression::LambdaExpression(_, b)) => a == b,
+            (Expression::ArrayLiteral(_, a), Expression::ArrayLiteral(_, b)) => a == b,
+            (Expression::BinaryOperation(_, a), Expression::BinaryOperation(_, b)) => a == b,
+            (Expression::UnaryOperation(_, a), Expression::UnaryOperation(_, b)) => a == b,
+            (Expression::IndexAccess(_, a), Expression::IndexAccess(_, b)) => a == b,
+            (Expression::FunctionCall(_, a), Expression::FunctionCall(_, b)) => a == b,
+            (Expression::FreeInput(_, a), Expression::FreeInput(_, b)) => a == b,
+            (Expression::MatchExpression(_, a), Expression::MatchExpression(_, b)) => a == b,
+            (Expression::IfExpression(_, a), Expression::IfExpression(_, b)) => a == b,
+            (Expression::BlockExpression(_, a), Expression::BlockExpression(_, b)) => a == b,
+            _ => false,
```

Comment:
I think this could be quite dangerous. If we add something to Expression, we will probably not extend this implementation.

Is there a different way of doing this? We probably don't want to clone and set the source ref, but maybe some simple macro?

Reply:
I ended up implementing all the comparison inside a macro and using all the variants to make the match exhaustive (without using wildcards), in case we add a new variant to Expressions, the match will complain.

What do you think?